docs/hazmat/primitives/smime.rst

```diff
@@ -0,0 +1,126 @@
+.. class:: SMIMESignatureBuilder
+
+    .. versionadded:: 3.2
+
+    .. doctest::
+
+        >>> from cryptography.hazmat.primitives import hashes, serialization, smime
+        >>> from cryptography import x509
+        >>> cert = x509.load_pem_x509_certificate(ca_cert)
+        >>> key = serialization.load_pem_private_key(ca_key, None)
+        >>> options = [smime.SMIMEOptions.DetachedSignature]
+        >>> smime.SMIMESignatureBuilder().set_data(
+        ...     b"data to sign"
+        ... ).add_signer(
+        ...     cert, key, hashes.SHA256()
+        ... ).sign(
+        ...     serialization.Encoding.PEM, options
+        ... )
+        b'...'
+
+    .. method:: set_data(data)
+
+        :param data: The data to be hashed and signed.
+        :type data: :term:`bytes-like`
+
+    .. method:: add_signer(certificate, private_key, hash_algorithm)
+
+        :param certificate: The :class:`~cryptography.x509.Certificate`.
+
+        :param private_key: The
+            :class:`~cryptography.hazmat.primitives.asymmetric.rsa.RSAPrivateKey` or
+            :class:`~cryptography.hazmat.primitives.asymmetric.ec.EllipticCurvePrivateKey`
+            associated with the certificate provided.
+
+        :param hash_algorithm: The
+            :class:`~cryptography.hazmat.primitives.hashes.HashAlgorithm` that
+            will be used to generate the signature. This must be an instance of
+            :class:`~cryptography.hazmat.primitives.hashes.SHA1`,
+            :class:`~cryptography.hazmat.primitives.hashes.SHA224`,
+            :class:`~cryptography.hazmat.primitives.hashes.SHA256`,
+            :class:`~cryptography.hazmat.primitives.hashes.SHA384`, or
+            :class:`~cryptography.hazmat.primitives.hashes.SHA512`.
+
+    .. method:: sign(encoding, options, backend=None)
+
+        :param encoding: :attr:`~cryptography.hazmat.primitives.serialization.Encoding.PEM`
+            or :attr:`~cryptography.hazmat.primitives.serialization.Encoding.DER`.
+
+        :param options: A list of :class:`~cryptography.hazmat.primitives.smime.SMIMEOptions`.
+
+        :param backend: An optional backend.
+
+
+.. class:: SMIMEOptions
+
+    .. versionadded:: 3.2
+
+    An enumeration of options for S/MIME signature creation.
+
+    .. attribute:: Text
+
+        The text option adds ``text/plain`` headers to the S/MIME message when
+        serializing to
+        :attr:`~cryptography.hazmat.primitives.serialization.Encoding.PEM`.
+        This option has no effect with ``DER`` serialization.
+
+    .. attribute:: Binary
+
+        S/MIME signing normally converts line endings (LF to CRLF). When
+        passing this option the data will not be converted.
+
+    .. attribute:: DetachedSignature
+
+        Don't embed the signed data within the ASN.1. When signing with
+        :attr:`~cryptography.hazmat.primitives.serialization.Encoding.PEM` this
+        also results in the data being added as clear text before the
+        PEM encoded structure.
```

Comment:
Wait what? So if it's DER then this option means "sig only, no signed data" but for PEM this means "signed data moved somewhere else"? That's not intuitive.

Reply:
Nope, it's not.

Reply:
Can we only support this option with DER then? That's why too confusing to reason about.

Reply:
No, we have consumers in the issue that directly depend on the detached PEM output

Reply:
If they were two different options, do you think we could document this more coherently?

Reply:
Do you have a proposal for what those options would look like? We need to be able to clearly articulate what set of options users coming from OpenSSL need to use, which is challenging since most of them are likely unaware of the difference between PEM and DER here (most S/MIME is done PEM and DER is just the PKCS7 structure sans the additional S/MIME trappings).

Reply:
This weirdness seems pretty prevalent in online docs about S/MIME. We could call the S/MIME PEM detached sig form "Cleartext signing" but it's pretty consistently called detached signature (as is the concept of an ASN.1 PKCS7 payload with no embedded data).

Reply:
Ok. I guess that's it then.